### PR TITLE
[manuf] update perso flow to validate the DICE chain with OpenSSL on the host

### LINKED ISSUE #24954: [manuf] update FT provisioning flow to check DICE cert chain validates with OpenSSL
Currently, all host endorsed certs (except the UDS DICE cert) are [validated](https://github.com/lowRISC/opentitan/blob/a18eb97ad2275f8b0a62e9e969c2431d10b1e14f/sw/host/provisioning/ft_lib/src/lib.rs#L371) with openssl as a sanity check during the FT provisioning flow. This was because the DICE certs use an X.509 critical custom extension (DiceTcbInfo) that is unsupported by OpenSSL. However, the `-ignore_critical` [flag](https://docs.openssl.org/3.0/man1/openssl-verification-options/#verification-options) can be used to still check the UDS cert validates properly. The FT provisioning flow should be updated to:
- [x] check the UDS cert validates with OpenSSL
- [x] check the entire DICE chain validates with OpenSSL.

### PR DESCRIPTION
**Pull request by timothytrippel**
This updates the perso flow to verify the entire DICE certificate chain using OpenSSL on the host side. This fixes #24954.

### CODE REVIEW

**Review by vbendeb — COMMENTED** [review]
I started the review and midway a github message popped up about a patchset update and refusing to accept any more comments. Let me try publishing what I have so far

File: sw/host/provisioning/ft_lib/src/lib.rs [hunk 1]
```diff
@@ -423,9 +443,14 @@ fn provision_certificates(
...
 
     // Validate the certificate endorsements with OpenSSL.
-    if !host_endorsed_certs.is_empty() {
-        validate_certs_chain(ca_certificate.to_str().unwrap(), &host_endorsed_certs)?;
+    log::info!("Validating DICE certificate chain with OpenSSL ...");
+    validate_cert_chain(ca_certificate.to_str().unwrap(), &dice_cert_chain, true)?;
+    log::info!("Success.");
+    log::info!("Validating SKU-specific certificates with OpenSSL ...");
+    if !sku_specific_certs.is_empty() {
+        validate_certs(ca_certificate.to_str().unwrap(), &sku_specific_certs)?;
```
Comment (vbendeb) [inline sw/host/provisioning/ft_lib/src/lib.rs:451]: why do you need both `validate_certs()` and `validate_cert_chain()` 

You could create a vector of two certs here, replacing the second element with one SKU cert at a time and invoke `validate_certs()` on this vector in the loop

**Review by vbendeb — APPROVED** [review]

**Review by vbendeb — APPROVED** [review]

**Review by moidx — APPROVED** [review]

**Review by vbendeb — COMMENTED** [review]
this could have been squashed with the last patch in this chain...

File: sw/host/provisioning/cert_lib/src/lib.rs [hunk 2]
```diff
@@ -151,64 +151,129 @@ fn parse_and_endorse_x509_cert_ckms(tbs: Vec<u8>, ckms_key_id: &str) -> Result<V
...
+        // Overwrite the current leaf cert PEM file.
+        tmp_leaf_cert_pem_filename = write_cert_to_temp_pem_file(&cert.bytes, "leaf")?;
+
+        // Verify the cert chain up to the current leaf cert.
+        let mut args = vec!["verify", "-CAfile", tmp_ca_pem_filename];
+        if ignore_critical {
+            args.push("-ignore_critical");
+        }
+        args.push(tmp_leaf_cert_pem_filename.as_str());
+        openssl_command(&args).context("failed to verify a certificate chain")?;
```
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/host/provisioning/cert_lib/src/lib.rs:265]: Fixed.

File: sw/host/provisioning/cert_lib/src/lib.rs [hunk 3]
```diff
@@ -151,64 +151,129 @@ fn parse_and_endorse_x509_cert_ckms(tbs: Vec<u8>, ckms_key_id: &str) -> Result<V
...
+        // Cleanup the temp PEM cert file.
+        fs::remove_file(pem_filename.as_str()).context("failed to remove PEM file")?;
     }
 
-    fs::remove_file(der_filename).context("failed to remove der file")?;
-    fs::remove_file(pem_filename).context("failed to remove pem file")?;
+    Ok(())
+}
+
+/// Validate a chain of X.509 certificates against a provide CA certificate.
```
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/host/provisioning/cert_lib/src/lib.rs]: Fixed.

File: sw/host/provisioning/cert_lib/src/lib.rs [hunk 4]
```diff
@@ -151,64 +151,129 @@ fn parse_and_endorse_x509_cert_ckms(tbs: Vec<u8>, ckms_key_id: &str) -> Result<V
...
+    let pem_filename = binding_pem.as_str();
+
+    // Write DER bytes to the tmp file.
+    let size = get_cert_size(der_cert_bytes)?;
+    let mut file = OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .create(true)
+        .open(der_filename)
+        .context("failed to open temporary der file")?;
```
[the remark this answers is not in the document]
Comment (timothytrippel) [reply sw/host/provisioning/cert_lib/src/lib.rs]: Fixed.

**Comment by timothytrippel** [comment]
> this could have been squashed with the last patch in this chain...

Done.

File: sw/host/provisioning/ft_lib/src/lib.rs (hunk 1 above)
Comment (timothytrippel) [reply sw/host/provisioning/ft_lib/src/lib.rs]: good point. Fixed.

**Review by vbendeb — APPROVED** [review]

**Review by vbendeb — APPROVED** [review]

File: sw/device/silicon_creator/manuf/base/ft_personalize.c [hunk 5]
```diff
@@ -446,7 +448,9 @@ static status_t personalize_gen_dice_certificates(ujson_t *uj) {
                             &cdi_0_key_ids, &curr_pubkey, all_certs,
                             &curr_cert_size));
   cdi_0_offset = perso_blob_to_host.next_free;
-  TRY(perso_tlv_prepare_cert_for_shipping("CDI_O", false, all_certs,
+  // DO NOT CHANGE THE "CDI_0" STRING BELOW with modifying the `dice_cert_names`
+  // collection in sw/host/provisioning/ft_lib/src/lib.rs.
+  TRY(perso_tlv_prepare_cert_for_shipping("CDI_0", false, all_certs,
```
Comment (vbendeb) [inline sw/device/silicon_creator/manuf/base/ft_personalize.c:453]: I meant to squash first and last patches of this stack  actually, but nevermind :)

**Review by vbendeb — APPROVED** [review]
presumably the SKU tests pass :)

**Pull request merged**